### DISCUSSION TITLE
High-availability deployments

### DISCUSSION THREAD

**Discussion by ikaruswill** [umputun/remark42#1135] [General]
Hi, first off, I'd like to say, this is a very well written web app, thanks for taking the time to build and continuously improve this amidst so many support requests that you guys are responding out of your own free time.

I'm just starting to look at alternatives to Commento, which currently appears to be unmaintained and found Remark42. I've since built an arm64 image at [ikaruswill/remark42](https://hub.docker.com/repository/docker/ikaruswill/remark42) (Open for public use, happy to contribute) and test deployments are working well.

However, I do wonder, how does one deploy more than 1 instance of remark42? I noticed the use of a pub-sub cache, so in theory, there is _some_ support for multi-instance deployments, but since the database is local and file-based, I'm not sure. I did not see any documentation on high-availability deployments, so I figured I'll reach out here.

**Comment by umputun** [comment]
> I'm just starting to look at alternatives to Commento, which currently appears to be unmaintained and found Remark42. I've since built an arm64 image at ikaruswill/remark42 (Open for public use, happy to contribute) and test deployments are working well.

I'm not sure what this image does.  To me, it looks like the normal remrak42 just with a different tag (name). Is it some custom version of the remark42 build or some private/custom code involved?

> However, I do wonder, how does one deploy more than 1 instance of remark42? I noticed the use of a pub-sub cache, so in theory, there is some support for multi-instance deployments, but since the database is local and file-based, I'm not sure. I did not see any documentation on high-availability deployments, so I figured I'll reach out here.

I have answered a similar question [here](https://github.com/umputun/remark42/discussions/1112). Generally speaking, there are two parts that should be done: add a custom data store and switch to an external cache. The auth should be fine as-is due to JWT nature. Theoretically, you may even turn the caching off if your datastore is fast or provides sufficient caching by itself.

> I did not see any documentation on high-availability deployments

This is because we don't provide any high-availability deployments options, but rather allow users to make their own. For end-users, remark42 is not HA-ready out of the box, and the only way to achieve it is by adding things I mentioned above. I'm not sure how this can be documented and to me, the code itself is enough documentation. However, if you going to  implement HA version it would be very nice if you can document the process and provide a PR with this documentation.

**Comment by ikaruswill** [comment]
I think you might have misunderstood my first point. 

That's no custom image, but rather a multiarch build of your image. I'm just putting it out here to share in case anyone is looking for an arm64/amd64 image for use in a hybrid arch cluster.

Roger that thanks for the link. I just wanted to know if anyone has already figured it out. And yes don't get me wrong, I'm not saying that the documentation is inadequate or the code doesn't speak for itself, all I wanted is to find out if I missed something somewhere or if someone's already deployed HA remark42.

Thanks for the quick reply nonetheless.

**Reply by umputun** [reply]
got it. 

btw, thx for the reminder. I have added multi-arch build to the master and it pushes images to both docker hub and [ghcr](https://github.com/umputun/remark42/pkgs/container/remark42).

**Reply by ikaruswill** [reply]
That's great, thanks for putting up the multiarch builds. Appreciate your work sir. Have a good one.
